proxy/src/com/aerospike/client/proxy/BatchProxy.java

```diff
@@ -0,0 +1,138 @@
+        private final Key[] keys;
+        private final Operation[] ops;
+        private final BatchAttr attr;
+        private final boolean isOperation;
+
+        public OperateRecordSequenceCommandProxy(
+                GrpcCallExecutor grpcCallExecutor,
+                BatchPolicy batchPolicy,
+                Key[] keys,
+                Operation[] ops,
+                BatchRecordSequenceListener listener,
+                BatchAttr attr
+        ) {
+            super(grpcCallExecutor, batchPolicy);
+            this.batchPolicy = batchPolicy;
+            this.keys = keys;
+            this.ops = ops;
+            this.sent = new boolean[keys.length];
+            this.listener = listener;
+            this.attr = attr;
+            this.isOperation = ops != null;
+        }
+
+        @Override
+        protected MethodDescriptor<Kvs.AerospikeRequestPayload, Kvs.AerospikeResponsePayload> getGrpcMethod() {
+            return KVSGrpc.getBatchOperateStreamingMethod();
+        }
+
+        @Override
+        protected boolean isUnaryCall() {
+            return false;
+        }
+
+        @Override
+        void writeCommand(Command command) {
+            // The destination node is a single Aerospike proxy instance,
+            // where all keys are sent to the same node. Keys are not
+            // distributed across nodes.
+            // TODO @BrianNichols can a interface be implemented by both the
+            //  native and proxy client, to be passed to setBatchOperate as an
+            //  argument? node is passed in as "null" in BatchNode constructor.
+            BatchNode batchNode = new BatchNode(null, keys.length, 0);
+            for (int i = 1; i < keys.length; i++) {
+                batchNode.addKey(i);
+            }
+
+            command.setBatchOperate(batchPolicy, keys, batchNode, null, ops, attr);
+        }
+
+        @Override
+        void onResponse(Kvs.AerospikeResponsePayload response) {
+            byte[] bytes = response.getPayload().toByteArray();
+            Parser parser = new Parser(bytes, response.getStatus());
+
+            if (response.getHasNext()) {
+                parseResult(parser, response.getInDoubt());
+                return;
+            }
+
+            int resultCode = parser.parseHeader(5);
+            if (resultCode == ResultCode.OK) {
+                onSuccess();
+            } else {
+                onFailure(new AerospikeException(resultCode));
+            }
+        }
+
+        @Override
+        void parseResult(Parser parser, Boolean inDoubt) {
+            int resultCode = parser.parseHeader(5);
+            parser.skipKey();
+
+            Key keyOrig = keys[parser.batchIndex];
+            BatchRecord record;
+
+            if (resultCode == 0) {
+                record = new BatchRecord(keyOrig, parser.parseRecord(isOperation), attr.hasWrite);
+            } else {
+                // TODO @BrianNichols commandSentCounter?
```

Comment:
How to handle commandSentCounter?

Reply:
commandSentCounter is not used on proxy client side.